src/domain_extractor.coffee

```diff
@@ -0,0 +1,111 @@
+    Then, starting from right to left we consider anything that is less than
+    4 characters long as part of the TLD. When we reach a level that is
+    at least 4 characters we include it in the domain and we stop.
+
+    We also accept a custom TLD list that allows us to bypass this.
+    In that case, the domain is determined as the custom TLD plus
+    the part immediately preceding it.
+
+    Finally, we always drop the www prefix.
+
+    This approach might be problematic in the future, but the only safe
+    alternative is to use the Public Suffix List which contains all
+    the known TLDs. However is quite large in size.
+
+    @see https://github.com/wrangr/psl
+  ###
+  class DomainExtractor
+    NUMBER_DOT_ONLY_PATTERN = /^[\d.]+$/
+
+    constructor: (hostname) ->
+      @hostname = hostname.toLowerCase();
+
+    ###
+      Retrieve the base domain from the given hostname
+
+      @param [Boolean] wildcard Whether to append a dot in the beginning of the domain.
+      @return [String, null] The base domain or null if is IP or 1 level
+    ###
+    get: (wildcard = true) ->
+      return null if @hostname.match NUMBER_DOT_ONLY_PATTERN # check if it is IP
+
+      @domainParts = @hostname.split('.');
+      @domainParts.shift() if @domainParts[0] == 'www' # remove the www prefix
+      return null if @domainParts.length == 1 # check if it is 1 level
+
+      matched_tld = @_matched_custom_tld()
+      domain = if matched_tld then @_construct_matched_domain(matched_tld) else @_construct_domain()
+      domain = '.' + domain if wildcard
+
+      domain
+
+    ###
+      Construct the domain name from the domain parts
+
+      @return [String] The domain
+    ###
+    _construct_domain: ->
+      domain = [@domainParts.pop()]
+
+      while @domainParts.length
+        domain.unshift(@domainParts.pop())
+        break if domain[0].length > 3
+
+      domain.join('.')
+
+    ###
+      Construct the domain based on the custom tld that was matched in the hostname
+      Keep the custom tld and extract the next domain levels if necessary
+
+      @param [String] matched_tld The custom tld that was matched in the hostname
+      @return [String] The domain
+    ###
+    _construct_matched_domain: (matched_tld) ->
+      matched_tld_length = matched_tld.split('.').length
+      start = @domainParts.length - matched_tld_length
+
+      # if another subdomain level exists, included it the domain
+      if start > 0
+        --start
+        ++matched_tld_length
+
+      @domainParts.splice(start, matched_tld_length).join('.')
+
+    ###
+      Check if the given hostname ends with one of the custom ltd we have set
+
+      @return [String, null] The matched tld or null if not matched
+    ###
+    _matched_custom_tld: ->
+      matched_tld = null
+
+      for tld in @_custom_tlds()
+        dotted_tld = '.' + tld
+        if @hostname.indexOf(dotted_tld, @hostname.length - dotted_tld.length) != -1
+          matched_tld = tld
```

Comment:
Why not return the `tld` here and omit the `matched_tld` variable?

Reply:
nice 👍